// jotai/umd/query.development.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tanstack/query-core'), require('jotai')) :
  typeof define === 'function' && define.amd ? define(['exports', '@tanstack/query-core', 'jotai'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.jotaiQuery = {}, global.queryCore, global.jotai));
})(this, (function (exports, queryCore, jotai) { 'use strict';

  var queryClientAtom = jotai.atom(new queryCore.QueryClient());

  function atomWithQuery(createQuery, getQueryClient) {
    if (getQueryClient === void 0) {
      getQueryClient = function getQueryClient(get) {
        return get(queryClientAtom);
      };
    }

    var previousDataCache = new WeakMap();
    var queryDataAtom = jotai.atom(function (get) {
      var queryClient = getQueryClient(get);
      var options = typeof createQuery === 'function' ? createQuery(get) : createQuery;
      var observer = new queryCore.QueryObserver(queryClient, options);
      var initialResult = observer.getCurrentResult();

      if (initialResult.data === undefined && options.keepPreviousData && previousDataCache.has(queryClient)) {
        initialResult.data = previousDataCache.get(queryClient);
      }

      var resolve = null;

      var makePending = function makePending() {
        return new Promise(function (r) {
          resolve = r;
        });
      };

      var resultAtom = jotai.atom(initialResult.data === undefined && options.enabled !== false ? makePending() : initialResult);
      var setResult = null;

      var listener = function listener(result) {
        if (result.isFetching || !result.isError && result.data === undefined) {
          return;
        }

        if (!resolve && !setResult) {
          throw new Error('setting result without mount');
        }

        if (resolve) {
          resolve(result);
          resolve = null;
        }

        if (setResult) {
          setResult(result);
        }

        if (result.data !== undefined) {
          previousDataCache.set(queryClient, result.data);
        }
      };

      var unsubscribe = null;
      var timer;

      var startQuery = function startQuery() {
        if (!setResult && unsubscribe) {
          clearTimeout(timer);
          unsubscribe();
          unsubscribe = null;
        }

        if (options.enabled !== false) {
          if (setResult) {
            observer.refetch({
              cancelRefetch: true
            }).then(setResult);
          } else {
            unsubscribe = observer.subscribe(listener);
          }
        }

        if (!setResult) {
          timer = setTimeout(function () {
            if (unsubscribe) {
              unsubscribe();
              unsubscribe = null;
            }
          }, 1000);
        }
      };

      startQuery();

      resultAtom.onMount = function (update) {
        setResult = update;

        if (unsubscribe) {
          clearTimeout(timer);
        } else {
          startQuery();
        }

        return function () {
          setResult = null;

          if (unsubscribe) {
            unsubscribe();
          }
        };
      };

      return {
        resultAtom: resultAtom,
        makePending: makePending,
        startQuery: startQuery
      };
    });
    var queryAtom = jotai.atom(function (get) {
      var _get = get(queryDataAtom),
          resultAtom = _get.resultAtom;

      var result = get(resultAtom);

      if (result.isError) {
        throw result.error;
      }

      return result.data;
    }, function (get, set, action) {
      var _get2 = get(queryDataAtom),
          resultAtom = _get2.resultAtom,
          makePending = _get2.makePending,
          startQuery = _get2.startQuery;

      switch (action.type) {
        case 'refetch':
          {
            set(resultAtom, makePending());
            startQuery();
          }
      }
    });
    return queryAtom;
  }

  function atomWithInfiniteQuery(createQuery, getQueryClient) {
    if (getQueryClient === void 0) {
      getQueryClient = function getQueryClient(get) {
        return get(queryClientAtom);
      };
    }

    var queryDataAtom = jotai.atom(function (get) {
      var queryClient = getQueryClient(get);
      var options = typeof createQuery === 'function' ? createQuery(get) : createQuery;
      var defaultedOptions = queryClient.defaultQueryOptions(options);
      var observer = new queryCore.InfiniteQueryObserver(queryClient, defaultedOptions);
      var initialResult = observer.getCurrentResult();
      var resolve = null;
      var resultAtom = jotai.atom(initialResult.data === undefined && options.enabled !== false ? new Promise(function (r) {
        resolve = r;
      }) : initialResult);

      var setResult = function setResult() {
        throw new Error('setting result without mount');
      };

      var state = {
        isMounted: false,
        unsubscribe: null
      };

      var listener = function listener(result) {
        if (result.isFetching || !result.isError && result.data === undefined || result.isError && queryCore.isCancelledError(result.error)) {
          return;
        }

        if (resolve) {
          setTimeout(function () {
            if (!state.isMounted) {
              state.unsubscribe == null ? void 0 : state.unsubscribe();
              state.unsubscribe = null;
            }
          }, 1000);
          resolve(result);
          resolve = null;
        } else {
          setResult(result);
        }
      };

      if (options.enabled !== false) {
        state.unsubscribe = observer.subscribe(listener);
      }

      resultAtom.onMount = function (update) {
        setResult = update;
        state.isMounted = true;

        if (options.enabled !== false && !state.unsubscribe) {
          state.unsubscribe = observer.subscribe(listener);
          listener(observer.getCurrentResult());
        }

        return function () {
          return state.unsubscribe == null ? void 0 : state.unsubscribe();
        };
      };

      return {
        options: options,
        resultAtom: resultAtom,
        observer: observer,
        state: state
      };
    }, function (get, set, action) {
      var _get2 = get(queryDataAtom),
          options = _get2.options,
          resultAtom = _get2.resultAtom,
          observer = _get2.observer,
          state = _get2.state;

      if (options.enabled === false) {
        return;
      }

      switch (action.type) {
        case 'refetch':
          {
            set(resultAtom, new Promise(function () {}));

            if (!state.isMounted) {
              state.unsubscribe == null ? void 0 : state.unsubscribe();
              state.unsubscribe = null;
            }

            observer.refetch(action.payload).then(function (result) {
              set(resultAtom, result);
            });
            return;
          }

        case 'fetchPreviousPage':
          {
            observer.fetchPreviousPage();
            return;
          }

        case 'fetchNextPage':
          {
            observer.fetchNextPage();
            return;
          }
      }
    });
    var queryAtom = jotai.atom(function (get) {
      var _get3 = get(queryDataAtom),
          resultAtom = _get3.resultAtom;

      var result = get(resultAtom);

      if (result.isError) {
        throw result.error;
      }

      return result.data;
    }, function (_get, set, action) {
      return set(queryDataAtom, action);
    });
    return queryAtom;
  }

  exports.atomWithInfiniteQuery = atomWithInfiniteQuery;
  exports.atomWithQuery = atomWithQuery;
  exports.queryClientAtom = queryClientAtom;

  Object.defineProperty(exports, '__esModule', { value: true });

}));
